File: 현세미/프로그래머스/실패율.py
# 프로그래머스 - 실패율
def solution(N, stages):
    stage_counts = [0] * (N + 2)
    for stage in stages:
        stage_counts[stage] += 1

    total = len(stages)
    ratio = []
    for i in range(1, N + 1):
        if total > 0:
            fail_rate = stage_counts[i] / total
            ratio.append((i, fail_rate))
            total -= stage_counts[i]
        else:
            ratio.append((i, 0))

    ratio.sort(key=lambda x: (-x[1], x[0]))
    return [stage[0] for stage in ratio]

File: 이진영/프로그래머스/실패율.js
function solution(N, stages) {
    const stageLog = Array(N + 1).fill({}).map((_) => {
        return {
            notClear: 0,
            total : 0,
        }
    });
    
    stages.forEach(v => {
        if (v <= N) {
            stageLog[v].notClear++;
        
            for (let i = 1; i <= v; i++) {
                stageLog[i].total++;
            }
        } else {
            for (let i = 1; i <= N; i++) {
                stageLog[i].total++;
            }
        }
    })
 
    return Array(N).fill(0).map((_, i) => i + 1)
        .sort(
            (a, b) =>
            (stageLog[b].notClear / stageLog[b].total) -
            (stageLog[a].notClear / stageLog[a].total) 
        )
}